### ISSUE TITLE
Checkbox value breaking pristine state

### ISSUE THREAD

**Issue by vitbokisch** [erikras/redux-form#2857]
Hi,
I found out that using true/false value within checkboxes is breaking the pristine state value. The case is: Checkbox initial value is probably an empty string. When I set a checkbox to checked state and then uncheck it, the pristine value changes. The flow is '' -> true -> false. You can even check it in the simple example here: http://redux-form.com/6.6.3/examples/simple/ 

I fixed it by using normalize prop in Field for checkbox component like: `<Field normalize={value => value || ''} ... />`. Another approach could be setting default initial values for all the checkboxes in each form, which doesn't make sense at all.

I propose to set as a default behavior true || '' for checkboxes. Then the checkboxes will behave the same way as any other types of inputs.

**Comment by erikras** [comment]
Published fix as [`v6.7.0`](https://github.com/erikras/redux-form/releases/tag/v6.7.0).

**Comment by dmeenhuis** [comment]
Not sure if intended, but this change now breaks the application I'm currently working on. I'm submitting boolean values from my form to an API which expects the relevant property to be an actual boolean value (``true``/``false``).

However, after this change, when I uncheck the box, it now submits an empty string while previously it submitted ``false``. The API cannot convert the empty string to a boolean, so I'm getting an error.

**Comment by vitbokisch** [comment]
@dmeenhuis if it worked before and you use Field component, you can just use the opposite approach I've been using (I mean using normalize). Could work like this:
```
<Field 
    component={Control} 
    name={name} 
    normalize={value => value || false}
    {...props} 
  /> 
```
Also using props `parse` or `format` may help.

**Comment by dmeenhuis** [comment]
Sure, there are plenty of workarounds to think of. I'm just not sure if the current default behaviour introduced by this change is really sensible.

It doesn't feel right that something that worked like it should before (checkbox checked means a ``true`` value, unchecked is a ``false`` value), now requires additional code to make it work logically again. It's also pretty easy to forget, causing any receiving system that expects an actual boolean value to break. Perhaps @erikras would like to chime in?

Is there any way we can get the change introduced by this PR to work together with the previous behaviour?

**Comment by potty** [comment]
Shouldn't be this change marked as BC break? Suddenly my app stopped working after update.

**Comment by geminiyellow** [comment]
This change is actually a fuck BC

**Comment by theKirt** [comment]
The solution posted by @vitbokisch works 👍

**Comment by kgram** [comment]
This seems like a very bad solution to the problem. If your checkbox is initialised with an empty string as the value, isn't the problem this initial wrong value rather than any subsequent correct values?

Since this breaking change was released as a minor version you probably very silently broke most applications using checkboxes in redux-form. It certainly broke mine.

**Comment by giedriusvickus** [comment]
This should be marked as breaking change. Also, i don't see how this could make sense. This does not brake pristine, if you would use it with right initial values. Checkbox is not an input, where you could/should have value as a string. I hope this will be fixed in future versions!

Reverting to 6.6.3

**Comment by naw** [comment]
This change moves the so-called problem from people who use `''` as the initial value to people who use `false` as the initial value.  For example, if your initial value is `false`, checking and then unchecking the checkbox moves the value from `false` to `true` to `''`, which presumably causes the same issue with `pristine` that the OP was concerned about.

Regardless of whether `''` is a better unchecked value than `false`, this is a _substantial_ change in behavior. Would you be willing to update the release notes for 6.7.0 to reflect this as a breaking change?

As for the underlying `value` issue, perhaps there is room to make redux-form configurable with `checkedValue` and `uncheckedValue` configurations to be used for all checkboxes?

**Comment by cristian-sima** [comment]
My opinion is that is a bad decizion as I have to add everywhere `normalize={value => value || false}`. 

My app is sending the form to the Golang server, which tries to read the JSON, which contains "". Therefore the app returns an error.

Any chance to see this problem improved?

**Comment by vitbokisch** [comment]
@kgram, @giedriusvickus : with your proposal, we should always define initial state for each checkbox in all forms in your app which does not make sense at all when you want to render plain default form.

@cristian-sima In case you have a long long list of checkboxes it does not make sense at all to store false value in redux. In redux, you should store as many information as needed, not more. Otherwise your app could be very slow. In case you had list of 1000 checkboxes before this change you stored all the false values in redux without any reason. 

ECMAScript language specification:

> The result is false if the argument is the empty String (its length is zero); otherwise the result is true

Another thing is checkbox in redux-form does not follow exactly the specification of checkbox defined in HTML Specification otherwise it would return array of values:

> Checkboxes (and radio buttons) are on/off switches that may be toggled by the user. A switch is “on” when the control element’s checked attribute is set. When a form is submitted, only “on” checkbox controls can become successful.
> Several checkboxes in a form may share the same control name. Thus, for example, checkboxes allow users to select several values for the same property. The INPUT element is used to create a checkbox control.

In another words this might speed up an application a bit and you do not spam your redux state with unnecessary information and it does not force you to set initial states without any reason (just to fix pristine state is not a reason).

If you guys need to change it so many times as you are saying, there might be a problem with your project structure as in my projects it is only one line change in whole project. ;)

_One of the solutions could be: normalize prop can have a default func prop normalize={value => value || false} But not sure if it's a must have. What do you think @erikras?_

**Comment by danielrob** [comment]
To me, an unchecked checkbox should either have it's inital falsey value, or not be a key in the store (as is the case when there is no initial value). Am I missing something?

**Comment by kgram** [comment]
@vitbokisch: I think you misunderstand why people are complaining as well as what we would prefer.

We're not complaining from aches in our wrists from all the excessive writing we had to do. While I hadn't wrapped the component, I only had to change something like 8 instances. This is not a matter of characters changed.

First of all, people are annoyed about a breaking change being released as a minor version. Maybe you work with a backend that treats `""` and `false` as the same and thus nothing happened. I don't. Frankly I prefer not to, as I wouldn't consider them the same. Everyone complaining came here looking for answers to an inexplicable bug that suddenly manifested. It took me hours to even find this thread and I imagine it's the same for many other people. How could I have avoided a suddenly broken app with a change in behaviour like this? Except of course by using shrinkwrap/lock-files, which I guess this whole thing is the perfect argument for.

Secondly, no one argues _for_ `""` -> `true` -> `false` being a good or logical pattern and no one wants pristine state to break for checkboxes. But to look at that trio and conclude `false` is the odd man out is baffling to me. The obvious pattern should be `false` -> `true` -> `false`. And no, this does not mean that you should have to initialise the value manually, the same logic resulting in the second `false` should result in the default value being `false`. Honestly, I think even `""` -> `"on"` -> `""` would be a better pattern than this change, at least it would be obvious what was happening and use a consistant type.

**Comment by danielrob** [comment]
I still stand by the flow should be: 
`undefined` -> `{ "employed": true }` -> `undefined`  if no initial values are set and 
if `initalValues` is `''`: 
`{ "employed": '' }` -> `{ "employed": true }` -> `{ "employed": '' }`
if `initalValues` is `false`: 
`{ "employed": false }` -> `{ "employed": true }` -> `{ "employed": false }`

Yes, at the DOM level, an attribute can only have a false value by having an empty string. But as @vitbokisch points out this is reconciled by the fact that in plain HTML forms, empty strings are ignored. In redux-form the equivalent of an ignored value is that it's key doesn't even exist in the form values.

**Comment by danielrob** [comment]
Taking a look at the code, I'm not sure my dream case above is easily achievable, and I'm otherwise in agreement with @kgram here...  @erikras would you care to explain what your thoughts were behind https://github.com/erikras/redux-form/pull/2863 making "perfect sense"?.

**Comment by kinyat** [comment]
@danielrob I still have `false` -> `true` -> `""` in version 7.0.1 though.

**Comment by erikras** [comment]
Published fix in [`v7.1.0`](https://github.com/erikras/redux-form/releases/tag/v7.1.0).
